add block data handler for single body block transmissions

mkicherer opened the pull request:
The pull request following the discussion in #1758.

I did not test all use cases yet as I guess that there are (QA) things I missed anyways.

mkicherer committed:
add block data handler for single body block transmissions

mrdeep1 commented on include/coap3/coap_net.h:
```diff
@@ -117,6 +117,28 @@ typedef void (*coap_pong_handler_t)(coap_session_t *session,
+ * @param session The CoAP session.
+ * @param pdu The received PDU.
+ * @param resource The resource that receives the data (is NULL for clients)
+ * @param body_data The libcoap internal pointer to the complete data
```
Please end parameter comments with a . (period).

mrdeep1 commented on include/coap3/coap_net.h:
```diff
@@ -117,6 +117,28 @@ typedef void (*coap_pong_handler_t)(coap_session_t *session,
+ * @param length    The length of @p data.
+ * @param data      The payload data of the current block.
+ * @param offset    The offset of the @p data block.
+ * @param total     The estimated total size of all blocks.
```
A @return is needed

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -137,6 +140,32 @@ typedef void (*coap_release_large_data_t)(coap_session_t *session,
+ * @param session The CoAP session.
+ * @param pdu The received PDU.
+ * @param resource The resource that receives the data (is NULL for clients)
+ * @param body_data The libcoap internal pointer to the complete data
```
Please end parameter comments with a . (period).

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -337,6 +367,31 @@ body dropped.
+COAP_BLOCK_SINGLE_BODY is enabled. This is especially relevant for cases where
+COAP_BLOCK_SINGLE_BODY is enabled implicitly, e.g., when COAP_BLOCK_TRY_Q_BLOCK
+is enabled, and the application wants to process the single data blocks
+nonetheless but without handling the complicated protocol signaling itself.
```
s/nonetheless/individually/

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -337,6 +367,31 @@ body dropped.
+set to NULL.
+
+*NOTE*: the block data handler will be silently ignored for (reverse) proxy
+resources.
```
*NOTE*: the call to the block data handler will be silently ignored for (forward | reverse) proxy
resources.

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -337,6 +367,31 @@ body dropped.
+if the single blocks are only of interest in specific cases. It can also be
+used to store a pointer to memory allocated by the application that will later
+be presented to the method or response handler after the transmission finished
+eventually.
```
be presented to the method or response handler after the complete data body has been transferred.

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -137,6 +141,32 @@ typedef void (*coap_release_large_data_t)(coap_session_t *session,
+ * @param data      The payload data of the current block.
+ * @param offset    The offset of the @p data block.
+ * @param total     The estimated total size of all blocks.
+ */
```
A @return is needed

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -583,6 +638,77 @@ main(int argc, char *argv[]) {
+  /* Initialize libcoap library */
+  coap_startup();
+
+  (void)argc;
+  (void)argv;
```
I would prefer it if these were moved to before coap_startup(); as done elsewhere.

mrdeep1 commented on include/coap3/coap_net.h:
```diff
@@ -117,6 +117,28 @@ typedef void (*coap_pong_handler_t)(coap_session_t *session,
+ * @param offset    The offset of the @p data block.
+ * @param total     The estimated total size of all blocks.
+ */
+typedef coap_pdu_code_t (*coap_block_data_handler_t)(coap_session_t *session,
```
I think this should be` typedef coap_response_t (.....`  and return codes are one of COAP_RESPONSE_FAIL or COAP_RESPONSE_OK.

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -583,6 +638,77 @@ main(int argc, char *argv[]) {
+  /* let libcoap gather all data blocks for client-side reception as example */
+  if (!resource) {
+    *body_data = coap_block_build_body(*body_data, length, data, offset, total);
+    return 0;
```
Correct return values.

mrdeep1 commented on src/coap_block.c:
```diff
@@ -3244,13 +3244,25 @@ coap_handle_request_put_block(coap_context_t *context,
+      resp_code = context->block_data_handler(session, pdu, resource, &lg_srcv->body_data,
+                                              length, data, saved_offset,
+                                              lg_srcv->total_len);
+      if (resp_code != COAP_EMPTY_CODE) {
+        response->code = resp_code;
+        goto skip_app_handler;
```
Using `coap_response_t`, I would send back the `COAP_RESPONSE_CODE(500);` here, rather than doing it in the application.

mrdeep1 commented on src/coap_block.c:
```diff
@@ -4121,10 +4139,20 @@ coap_handle_response_get_block(coap_context_t *context,
+              resp_code = context->block_data_handler(session, rcvd, 0,
+                                                      &lg_crcv->body_data, length,
+                                                      data, saved_offset, size2);
+              if (resp_code != 0) {
```
I think this should be `coap_response_t `here.

mrdeep1 reviewed (changes requested):
Thanks for doing this work.. In principle it is looking good - some comments to address.

anyc commented:
Ok, I applied all changes.

mrdeep1 commented on man/coap_block.txt.in:
```diff
@@ -583,6 +641,77 @@ main(int argc, char *argv[]) {
+     * In this example, we do not set body_data so the method handler will be
+     * called with data set to NULL and the method handler should then close the
+     * file handle as all data was received.
+     */
```
This code will not get executed on the second function call as it is protected by `if (!f) {`.

mkicherer commented:
Sorry, I was logged in with the wrong account and did not get the "mark resolved" button.

mrdeep1 reviewed (changes requested):
A couple of minor things to resolve.

I need to do a regression set of tests which may come up with a couple of other changes.

mrdeep1 commented:
I have found a couple of issues as well as added in some test capability in 
[patch.txt](https://github.com/user-attachments/files/22800012/patch.txt).

I am beginning to think your original idea of configuring this on/off per resource for the server is perhaps a better way to go, but still need equivalent logic for the client.  Having the server write everything out for every resource may be a bit of an overkill.

For the server, perhaps `coap_register_server_block_data_handler()` which has resource as the first parameter and for the client perhaps `coap_register_client_block_data_handler()` which has session as the first parameter.  As the second parameter can be NULL, this then gives the flexibility of turning on/off.

For example, a client can send information back and forward and then at the appropriate point call `coap_register_client_block_data_handler()` to cause subsequent data to be written out.

mkicherer commented:
I only added your fixes to the commit.

Maybe it would make sense to enable both methods? The current context-based approach and additionally a per-resource handler that overrides a possible context handler. Maybe it would be helpful to also make the other block flags selectable per resource? I am not sure.

For me, the context-based approach is sufficient as I just call `coap_block_build_body` for all cases except the file transfer.

mrdeep1 commented:
I think a way to go is to have a `coap_register_resource_block_data_handler()`, and then tag specific resource's requests to use this block_data_handler by setting COAP_RESOURCE_USE_BLOCK_DATA_HANDLER.

mkicherer commented:
Hm, the downside of a global handler registration I see is that I would have to store the resource-specific handler in my application's resource struct but I also see that this approach would be better for memory-constrained devices. Do you see this in addition to the context-based approach or would you prefer this in addition to another client-specific method?

mrdeep1 commented:
The registration using `coap_register_resource_block_data_handler()` could be done using the context or resource - if resource the handler would be kept in `coap_resource_t`.   If context, then in the `flags `parameter in` coap_resource_init() `needs `COAP_RESOURCE_USE_BLOCK_DATA_HANDLER `or'd in.

mkicherer commented:
I am currently debugging another issue in my application and I noticed that with your fixes from last week, the original request handler is not called anymore when the transfer is finished. Is this intentional or a side effect?

I am using the the block data handler to write the data in the file and the original method handler to close the file and start processing the file. If the original method handler is not called anymore, I also have to track if all blocks have been transmitted in my app in addition to libcoap.

mrdeep1 commented:
>  Is this intentional or a side effect?

Certainly not intentional.  However, as all the data has been received via `block_data_handler`, the presented PDU has no data which may be confusing your logic.  The only change relevant to the server was line 4270.

mkicherer commented:
I commented the two lines
```
pdu->used_size = pdu->data - pdu->token - 1;
pdu->data = NULL;
```
at line 3391 and the method handler is called again. I will try to understand this. Thank you!

mkicherer commented:
Ah, the method handler is called with these lines but `coap_get_data_large` returns zero for all values and my code does not continue in this case. Would it be possible to set pdu->data = NULL only if body_data is not set by the block_data_handler? So the method handler would be called with the final/total size.

mrdeep1 commented:
Hmm,  coap_get_data_large() at the start returns 
````
  *offset = pdu->body_offset;
  *total = pdu->body_total;
  if (pdu->body_data) {
    *data = pdu->body_data;
    *len = pdu->body_length;
    return 1;
  }
````
so commenting out the 2 lines at line 3391 makes no different here as the `pdu->data` check is after this code.

`block_data_handler `is only updating` lg_srcv->body_data`, and then a few lines before line 3391 `pdu `should be getting updated with the correct information.

So, something else is happening here.

mkicherer commented:
At least with my simple tests so far it would work this way. I will try to implement the suggested API for handler registration tomorrow. Thank you!

mkicherer commented:
My internal method handler is called now, this was an error on my side. I will have to change the example in the man page as block_data has to be set now (at least with an empty coap_binary_t) so the method handler, which closes the file, will be called.

> The registration using `coap_register_resource_block_data_handler()` could be done using the context or resource - if resource the handler would be kept in `coap_resource_t`. If context, then in the `flags `parameter in`coap_resource_init()`needs `COAP_RESOURCE_USE_BLOCK_DATA_HANDLER `or'd in.

A client would also register the handler with the context or would you like to see an additional `coap_register_session_block_data_handler` you mentioned in the beginning?

mkicherer commented:
I have updated the example code and added the `COAP_RESOURCE_USE_BLOCK_DATA_HANDLER` flag. Depending on your answer regarding the client side, I will also change the `coap_register_block_data_handler` function.

mrdeep1 commented:
Thanks for the update - I'll check it out and come back to you later when I have a few more spare cycles.

mrdeep1 commented:
> Depending on your answer regarding the client side, I will also change the coap_register_block_data_handler function.

I think we are good here.  Some suggested firther changes 
[patch.txt](https://github.com/user-attachments/files/22945871/patch.txt).

mkicherer commented:
Ok, I applied all changes. I assumed the changes examples/ and the tinydtls update were only for your local testing environment.

mrdeep1 commented:
Thanks for updating with the other changes.  I would prefer the changes to examples/ and appropriate man/ pages to be included as well (as demonstration code) so that I can then add in this `-3` parameter usage to my regression test suites (tinydtls is not a real change).

mkicherer commented:
Looks like one check is hanging again.

mrdeep1 commented:
Thanks for all the work done on coming up with this.  Just running a regression test which takes several hours before merging this PR.
